lib/Sema/CSBindings.cpp

```diff
@@ -25,6 +25,32 @@ ConstraintSystem::determineBestBindings() {
   // Look for potential type variable bindings.
   TypeVariableType *bestTypeVar = nullptr;
   PotentialBindings bestBindings;
+
+  // For purposes of best binding identification
+  // right-hand of `parameter binding` constraint
+  // is equivalent to the left-hand side, which
+  // represents declaration of the parameter and
+  // should drive the search.
+
+  SolverScope scope(*this);
+
+  for (auto &constraint : getConstraints()) {
+    if (constraint.getKind() != ConstraintKind::BindParam)
+      continue;
+
+    auto *lhs = constraint.getFirstType()->getAs<TypeVariableType>();
+    auto *rhs = constraint.getSecondType()->getAs<TypeVariableType>();
+
+    if (!(lhs && rhs))
+      continue;
+
+    auto rep1 = getRepresentative(lhs);
+    auto rep2 = getRepresentative(rhs);
+
+    if (rep1 != rep2)
+      mergeEquivalenceClasses(rep1, rep2, /*updateWorkList*/false);
+  }
+
```

Comment:
So the issue here is with the order in which we're attempting bindings. We shouldn't be attempting to propagate type information back from the parameter to the argument, but instead should be attempting bindings for `""` and then looking up the overloads, and using those to determine argument bindings.

This hack fixes this particular test case, but isn't a principled fix: https://gist.github.com/rudkx/b323a750686ada360dced7a5c2509c45

A more principled fix that would be more inline with how the solver works would be to add another field to `PotentialBindings` to track whether a type variable is the second type of a `BindParam` and these the last-ditch type variable to attempt to bind.